importer.py

```diff
@@ -587,16 +614,17 @@ def ensure_deck(col, note_name, deck_name):
         field_names = col.models.field_names(model)
 
         # Add missing fields at the end. Most notably, the Keisei one.
-        while len(field_names) < len(WKImporter.FIELDS):
-            col.models.add_field(model, col.models.new_field(WKImporter.FIELDS[len(field_names)]))
-            field_names = col.models.field_names(model)
-            ret = True
+        for field in WKImporter.FIELDS:
+            if field not in field_names:
+                col.models.add_field(model, col.models.new_field(field))
+                field_names = col.models.field_names(model)
+                ret = True
 
         if ret:
             col.models.update_dict(model)
             model = col.models.by_name(note_name)
 
-        if field_names != WKImporter.FIELDS:
+        if all(field in field_names for field in WKImporter.FIELDS) == False:
```

Comment:
This is a bad idea.
Anki-Imports are purely based on the order of Fields, not their names.
So if at all, this should truncate the field_names to the length of the FIELDS array, but then still take the order into account.

Reply:
Hm, I'm actually not sure anymore.
The NoteImporter is not exactly documented, and I can't tell from its code if the order is all that matters, or if it does look at dict keys like that.

Reply:
Actually, I was thinking this check can probably be removed all together since the logic before it now guaranties all missing fields are added to a card's model. So it should be impossible to hit.  

Also, I think maybe there is a misunderstanding of my logic in importer.py. When I generate the array of fields, I make sure that they are added in the correct order by using the existing card model as a guide. 

`self.col.models.field_names(self.model)` returns a list of the field names in order for our card model. I then iterate this to make sure that I am adding the fields to the card in the correct order. 

Also, I was just using this check to make sure all the correct fields just existed in our card. The purpose wasn't to check the order since my new logic doesn't care. I always make sure they are in the right order when generating the fields array for our ForeignNote.

I hope this makes sense.

Reply:
If you need a better explanation of my logic, I am more than happy to explain further to help you understand. And yes I agree, the API for anki is not very will documented. It took me quite a while to figure a lot of this out.

Reply:
The thing is, I'm not sure if the order matters after all.
Like, I'm looking at https://github.com/ankitects/anki/blob/0d8d816a0c424b1e033af7d4d1331816bbb61426/pylib/anki/importing/noteimp.py#L319
This function is what performs the actual updating of the fields on each card.
And it does so by iterating the self.mapping table, which is just a list, so `c` there is indeed an index.
And looking at https://github.com/ankitects/anki/blob/0d8d816a0c424b1e033af7d4d1331816bbb61426/pylib/anki/importing/noteimp.py#L99 this is generated from the actual field list in the cards, truncated or padded to the number returned from the actual model.

This gives me the idea that this whole code could be hugely simplified by also overriding the initMapping function, and altering its logic to replace fields the addon does not know about by `None`, which would make the logic in processFields skip that field and leaving the original value intact.
That would get rid of the entire block that fetches stuff from the database for each and every note.

Reply:
Or alternatively, after calling the original initMapping, iterate over the self.mapping list, and replace values not in the FIELDS list by `None`.

Reply:
That shouldn't be neccesary. The contents put into the field will be completely ignored if its entry in self.mappings is falsey. So it can be an empty string, None, or whatever.
Including those patterns in it is a neat idea. All it'd take would be to check if FETCH_CONTEXT_PATTERNS is set, and if not also None-Out its field in self.mappings.

I'd really like to avoid fetching data from the db inside there, and it does not seem neccesary according to the code of the NoteImporter.

Reply:
Alright, I will try to get to get to it soon and test. It might be today or tomorrow. Getting kind of tired now as its around 10pm where I live.